language/move-lang/src/expansion/dependency_ordering.rs

```diff
@@ -416,6 +493,36 @@ fn spec_block_member(context: &mut Context, sp!(_, sbm_): &E::SpecBlockMember) {
         M::Let { def: e, .. } | M::Include { exp: e, .. } | M::Apply { exp: e, .. } => {
             exp(context, e)
         }
-        M::Variable { .. } | M::Pragma { .. } => {}
+        // A special treatment to the `pragma friend` declarations.
+        //
+        // The `pragma friend = <address::module_name::function_name>` notion exists before the
+        // `friend` feature is implemented as a language feature. And it may still have a use case,
+        // that is, to friend a module that is compiled with other modules but not published.
+
+        // One example in the context of the Diem Framework is the `Genesis` module. As of now, the
+        // `Genesis` module is not published on chain. Therefore, we cannot really declare
+        // `friend Genesis;` in any module, because that will lead to linking error (the loader is
+        // unable to find `Genesis`). But the prover side still needs to know that `Genesis` is a
+        // friend (to verify global invariants). So, the `pragma friend = ...` syntax might need to
+        // stay. And for that, we need to add the module that is declared as a friend in the
+        // `immediate_neighbors`.
```

Comment:
Good explanation of whats going on! But, I'd reorient this as a general example and less Diem specific. There aren't really any Diem references in the move-lang crate, nor should their be

Reply:
Changed into a more general example.